client/landing/gutenboarding/stores/domain-suggestions/selectors.ts

```diff
@@ -14,13 +14,19 @@ import { stringifyDomainQueryObject } from './utils';
 export const getState = ( state: State ) => state;
 
 type DomainSuggestionSelectorOptions = Partial< Exclude< DomainSuggestionQuery, 'query' > >;
+
 export const getDomainSuggestions = (
 	state: State,
 	search: string,
 	options: DomainSuggestionSelectorOptions = {}
 ) => {
 	const normalizedQuery = normalizeDomainSuggestionQuery( search, options );
 
+	// If normalized search string (`query`) contains no alphanumerics, endpoint 404s
+	if ( ! normalizedQuery.query ) {
+		return [];
+	}
+
```

Comment:
With this in place, do we still need the [`?.` chaining in the component](https://github.com/Automattic/wp-calypso/pull/37595/files#diff-f8410a451fe06cb7cff0a51c9fecc4b2R53), and [`| undefined`](https://github.com/Automattic/wp-calypso/pull/37595/files#diff-f4b17720a150cc9d18fb0d903b7429b8R15) in the reducer type def?

Reply:
Ah yep, we do, since our state is keyed with queries and defaults to an empty object.

The only alternative would be to make the selector return an empty array if the key doesn't exists :thinking:

Reply:
E.g.

```diff
diff --git a/client/landing/gutenboarding/components/domain-picker/index.tsx b/client/landing/gutenboarding/components/domain-picker/index.tsx
index e5d61ef9af..0f84df6ba6 100644
--- a/client/landing/gutenboarding/components/domain-picker/index.tsx
+++ b/client/landing/gutenboarding/components/domain-picker/index.tsx
@@ -50,7 +50,7 @@ const DomainPicker: FunctionComponent = () => {
 						value={ domainSearch }
 					/>
 				</PanelRow>
-				{ suggestions?.length /* @TODO: Add optional chain */ ? (
+				{ suggestions.length /* @TODO: Add optional chain */ ? (
 					<PanelRow>
 						<ul>
 							{ suggestions.map( ( { domain_name } ) => (
diff --git a/client/landing/gutenboarding/stores/domain-suggestions/reducer.ts b/client/landing/gutenboarding/stores/domain-suggestions/reducer.ts
index d91a2535c8..059e516afc 100644
--- a/client/landing/gutenboarding/stores/domain-suggestions/reducer.ts
+++ b/client/landing/gutenboarding/stores/domain-suggestions/reducer.ts
@@ -12,7 +12,7 @@ import * as Actions from './actions';
 import { stringifyDomainQueryObject } from './utils';
 
 const domainSuggestions: Reducer<
-	Record< string, DomainSuggestion[] | undefined >,
+	Record< string, DomainSuggestion[] >,
 	ReturnType< typeof Actions[ 'receiveDomainSuggestions' ] >
 > = ( state = {}, action ) => {
 	if ( action.type === ActionType.RECEIVE_DOMAIN_SUGGESTIONS ) {
diff --git a/client/landing/gutenboarding/stores/domain-suggestions/selectors.ts b/client/landing/gutenboarding/stores/domain-suggestions/selectors.ts
index 0eb72f4b61..cfc8574e94 100644
--- a/client/landing/gutenboarding/stores/domain-suggestions/selectors.ts
+++ b/client/landing/gutenboarding/stores/domain-suggestions/selectors.ts
@@ -44,7 +44,7 @@ export const __internalGetDomainSuggestions = (
 	state: State,
 	queryObject: DomainSuggestionQuery
 ) => {
-	return state.domainSuggestions[ stringifyDomainQueryObject( queryObject ) ];
+	return state.domainSuggestions[ stringifyDomainQueryObject( queryObject ) ] ?? [];
 };
 
 /**
```

Reply:
I don't know, to be honest.

The selector is asking for an array of suggestions. Should it know that there's an HTTP request happening that will take an unknown amount of time to resolve and could fail?

There's a difference between `[]` (response of an empty list) and `undefined` - no response.

http-data makes that explicit:

https://github.com/Automattic/wp-calypso/blob/11e118d98184650504ad6afb453b112fb2e92162/client/state/data-layer/http-data.ts#L17-L41

I believe that's inspired by [this elm library](https://package.elm-lang.org/packages/krisajenkins/remotedata/latest/RemoteData#RemoteData) that also makes it explicit:

```elm
type RemoteData e a
    = NotAsked
    | Loading
    | Failure e
    | Success a
```

`undefined | SomeTypeOfResponse` let's us represent that we may have something or not, but its an ersatz way of representing something I'm not sure the selector needs to worry about. I believe wp-data exposes resolver state in some way through other APIs, but I'm not sure off the top of my head.

I'm inclined to leave _at least_ the `undefined` case. I tend to think it's better to have to deal with things (data may not be here yet) than to hide them away. We have the power to make that relatively straightforward to do with a decent type system.

Reply:
> [...]
> 
> `undefined | SomeTypeOfResponse` let's us represent that we may have something or not, but its an ersatz way of representing something I'm not sure the selector needs to worry about. I believe wp-data exposes resolver state in some way through other APIs, but I'm not sure off the top of my head.

Yeah, so all of my previous commentary was based on the assumption that we don't need to encode resolving/availability state into the return value of the selector. On a theoretical level, I find that to be a pattern that doesn't tend to scale well to all kinds of different possible return values and types. On a practical level, there's [`isResolving`](https://github.com/WordPress/gutenberg/blob/1e8d78189a9b05628b171016e102e2c50e267c39/packages/data/src/namespace-store/metadata/selectors.js#L55-L67) :grimacing: 

> I'm inclined to leave _at least_ the `undefined` case. I tend to think it's better to have to deal with things (data may not be here yet) than to hide them away. We have the power to make that relatively straightforward to do with a decent type system.

This is where I was torn. `undefined` seems to be a fairly generally applicable "hasn't been fetched"/"unavailable" marker, but I'm not quite sure about it in a world that's a) typed and b) tracks resolving state in some meta field, exposed through `isResolving`.

It'd basically mean that anything we'd fetch from the network will always be `| undefined`, and that we need to have guards against falsiness in consuming code. With resolving state tracked via `isResolving`, I find it very tempting to keep selector return types a bit more uniform -- e.g. always guarantee that we're getting an array back.

Reply:
👍 I'm not opposed to any of this, please go ahead.

Reply:
Okay, I had one more thought on this:

We need this check, to avoid probing the endpoint with what we know will return a 404. However, I think that the check should be colocated more closely to the `__internalGetDomainSuggestions` selector, since that selector could in theory also be called directly (even if we think of it as private, think of a hyptothetical situation where it was a different, public selector).

Now the next problem is that if we move it there, our `getDomainSuggestions` will call `select( STORE_KEY ).__internalGetDomainSuggestions( normalizedQuery )`, thus invoking the `__internalGetDomainSuggestions` _resolver_ -- the check in the `__internalGetDomainSuggestions` _selector_ can't prevent that.

Thus, I'm thinking the check should move into the `__internalGetDomainSuggestions` _resolver_ (bailing early, before the `apiFetch`) -- which means it'd propagate the fallback value (`[]`) through the entire action/reducer/selector chain.